File: guides/micronaut-jaxrs-jdbc/java/src/main/java/example/micronaut/resources/PetResource.java
package example.micronaut.resources;

import example.micronaut.domain.NameDto;
import example.micronaut.domain.Pet;
import example.micronaut.repositories.PetRepository;
import io.micronaut.data.model.Pageable;

import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import java.util.List;
import java.util.Optional;

@Path("/pets")
class PetResource {

    private final PetRepository petRepository;

    PetResource(PetRepository petRepository) {
        this.petRepository = petRepository;
    }

    @GET
    List<NameDto> all(Pageable pageable) {
        return petRepository.list(pageable);
    }

    @GET
    @Path("/{name}")
    Optional<Pet> byName(@PathParam("name") String petsName) {
        return petRepository.findByName(petsName);
    }

}

File: guides/micronaut-jaxrs-jdbc/python/src/example/micronaut/pet_resource.py
from typing import Annotated

from jakarta.validation import Valid
from jakarta.validation.constraints import NotNull
from jakarta.ws.rs import GET, POST, Path, PathParam
from micronaut.http import HttpStatus
from micronaut.http.annotation import Body, Status

from .name_dto import NameDto
from .pet import Pet
from .pet_repository import PetRepository
from .pet_save import PetSave


@Path("/pets")  # <1>
class PetResource:
    def __init__(self, pet_repository: PetRepository):  # <2>
        self.pet_repository = pet_repository

    @GET  # <3>
    def all(self) -> list[NameDto]:  # <4>
        return self.pet_repository.list()

    @GET  # <3>
    @Path("/{name}")  # <5>
    def by_name(self, pets_name: Annotated[str, PathParam("name")]) -> Pet | None:  # <6>
        return self.pet_repository.findByName(pets_name).orElse(None)

    @POST
    @Status(HttpStatus.CREATED)
    def save(self, pet_save: Annotated[PetSave, Body, NotNull, Valid]) -> None:
        self.pet_repository.save(pet_save.name, pet_save.type)
